torchgeo/datasets/geo.py

```diff
@@ -946,6 +1054,106 @@ def res(self, new_res: float) -> None:
         self.datasets[1].res = new_res
 
 
+class MultiQueryDataset(IntersectionDataset):
+    """Dataset representing a collection of Datasets.
+
+    This allows users to do things like:
+
+    * Spatio-temporal predictions where input sequences come from one dataset
+      and target sequences from another (e.g. Landsat and CDL)
+
+    These combinations require that all queries are present in *both* datasets,
+    and can be combined using an :class:`IntersectionDataset`:
+
+    This dataset should be used with :class:'SequentialGeoSampler' sampler.
+
+    .. versionadded:: 0.5.0
+    """
+
+    def __init__(
+        self,
+        input_dataset: GeoDataset,
+        target_dataset: GeoDataset,
+        transforms: Optional[Callable[[dict[str, Any]], dict[str, Any]]] = None,
+    ) -> None:
+        """Initialize a new Dataset instance.
+
+        Args:
+            input_dataset: the input sequence dataset
+            target_dataset: the target sequence dataset
+            transforms: a function/transform that takes input sample and its target as
+                entry and returns a transformed version
+
+        Raises:
+            ValueError: if either dataset is not a :class:`GeoDataset`
+        """
+        super().__init__(input_dataset, target_dataset, None, transforms)
+
+        self.input_dataset = input_dataset
+        self.target_dataset = target_dataset
+
+    def __getitem__(self, query: tuple[BoundingBox, BoundingBox]) -> dict[str, Any]:
+        """Retrieve image and metadata indexed by query.
+
+        Args:
+            query: two (minx, maxx, miny, maxy, mint, maxt) coordinates
+                to index datasets, where first indexes the input dataset
+                and second indexes the target dataset
+
+        Returns:
+            sample of data/targets and metadata at that index
+
+        Raises:
+            IndexError: if query is not within bounds of the index
+        """
+        if not query[0].intersects(self.bounds):
+            raise IndexError(
+                f"Input query: {query[0]} not found in input dataset "
+                "index with bounds: {self.bounds}"
+            )
+
+        if not query[1].intersects(self.bounds):
+            raise IndexError(
+                f"Target query: {query[1]} not found in target dataset "
+                "index with bounds: {self.bounds}"
+            )
+        # time-series where each query is for a different dataset
+        # assuming 1-to-1 correspondence between query order and dataset order
+        input_samples = self.input_dataset[query[0]]
+        target_samples = self.target_dataset[query[1]]
+
+        if self.transforms is not None:
+            input_samples = self.transforms(input_samples)
+            target_samples = self.transforms(target_samples)
+
+        samples = {
```

Comment:
Currently, this is intended with Time Series data as we are also returning the dates for example. I am not sure what other cases a more general `TupleDataset` would require and also how to handle possibly overlapping dictionary keys from the individual samples that you each retrieve, would one enumerate them for example? Then again I think it is nice to have the clarity betwenn "input" and "target" for time series tasks for example.